### PR TITLE
Simple coulombic interaction

### PR DESCRIPTION
**Pull request by sitangshugk95**
Incomplete, WIP. Don't merge yet

### CODE REVIEW

**Review by sitangshugk95 — COMMENTED** [review]
This is my first pull request, so I started with something relatively simple. Implements a simple Coulombic interaction for a given potential difference applied across capacitors of given thicknesses and permittivities as in Persson, B.N.J., 2018. The dependency of adhesion and friction on electrostatic attraction. The Journal of chemical physics, 148(14), p.144701. (sigma_zz(x) is the force, integrating which gives the work of electroadhesion, as in Heß, M. and Forsbach, F., 2020. Macroscopic modeling of fingerpad friction under electroadhesion: Possibilities and limitations. Frontiers in Mechanical Engineering, 6, p.567386. (Eq 22)

**Comment by sitangshugk95** [comment]
@sannant Please review this PR and merge if it looks correct. We can also discuss.

The next step would probably be to build a file that superimposes two interactions and returns an equivalent interaction that can be passed as an argument to make_system.
 
P.S. This is my first ever PR - is this the correct way to write new interactions/files and submit PRs?

File: Adhesion/Interactions/ES_Coulomb.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+#
+# Copyright 2020 Antoine Sanner
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:2]: We will need to rerun the updating of the license files

File: Adhesion/Interactions/ES_Coulomb.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+"""
+
+from Adhesion.Interactions import Potential
+
+import numpy as np
+from NuMPI import MPI
+
+
+class ES_C(Potential):
+    """ Electrostatic potential for Coulombic interaction when a voltage V is applied across an air gap between two perfect insulators of thicknesses d1,d2 and permittivities eps1,eps2
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:39]: Please add a reference to the paper here. Also consider using our @doi decorator to specify the doi of the paper making the derivation.

**Comment by sannant** [comment]
Can you please write one or a few tests asserting the code functionality at least ?  

At least like what I do here: 
https://github.com/ContactEngineering/Adhesion/blob/41fd4e8390d7c5f240a7f25c901516d2636aae47/test/Interactions/test_interactions.py#L1072

If there is reference data in the paper  or an analytic solution allowing to test the accuracy that would be perfect.

File: Adhesion/Interactions/ES_Coulomb.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            Thickness of top surface
+        V: float
+            Applied voltage
+        communicator: not used
+        """
+        self.eps1 = float(eps1)
+        self.eps2 = float(eps2)
+        self.d1 = float(d1)
+        self.d2 = float(d2)
+        self.V = float(V)
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:65]: call it voltage

File: Adhesion/Interactions/ES_Coulomb.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    def __repr__(self, ):
+        return ("Potential '{0.name}': eps1 = {0.eps1}, eps2 = {0.eps2}, d1 = {0.d1}, d2 = {0.d2}, V = {0.V},).format(self)
+
+    def evaluate(self, gap, potential=True, gradient=False, curvature=False,
+                 mask=None):
+        r = np.asarray(gap)
+
+        V = dV = ddV = None
+        
+        eps0 = 8.85e-12
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:86]: you can't hard code this value, it is not unitless. Set it as an attribute of the class at initialisation

File: Adhesion/Interactions/ES_Coulomb.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    """
+
+    name = "es-c"
+
+    def __init__(self, eps1, eps2, d1, d2, V, communicator=MPI.COMM_WORLD):
+        """
+        Parameters:
+        -----------
+        eps1: float
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:49]: These are the relative dielectric permittivities, correct ?

Please specify.

File: Adhesion/Interactions/ES_Coulomb.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                 mask=None):
+        r = np.asarray(gap)
+
+        V = dV = ddV = None
+        
+        eps0 = 8.85e-12
+	h0 = (self.d1 / self.eps1) + (self.d2 / self.eps2)
+	
+        if potential:
+            V = 0.5 * self.eps0 * (self.V**2) * (1 ./ (r + h0))
```
Comment (sannant) [inline Adhesion/Interactions/ES_Coulomb.py:90]: So the relative permittivity in the gap is 1, correct ? I would add it as a parameter still (with default value 1 if you whish)

File: Adhesion/Interactions/ES_Coulomb.py (hunk 5 above)
Comment (sitangshugk95) [reply Adhesion/Interactions/ES_Coulomb.py:51]: Yes, mentioned explicitly in definitions now

File: Adhesion/Interactions/ES_Coulomb.py (hunk 1 above)
Comment (sitangshugk95) [reply Adhesion/Interactions/ES_Coulomb.py:2]: Is this what you meant? I am not sure I understand what I should write in the first line and what the current license file number is.

File: Adhesion/Interactions/ES_Coulomb.py [hunk 7]
```diff
@@ -36,61 +36,67 @@
...
-    """ Electrostatic potential for Coulombic interaction when a voltage V is applied across an air gap between two perfect insulators of thicknesses d1,d2 and permittivities eps1,eps2
-
+    """ Electrostatic potential for Coulombic interaction when a voltage is applied across a gap between two perfect insulators
+@doi https://doi.org/10.3389/fmech.2020.567386
     """
 
     name = "es-c"
 
-    def __init__(self, eps1, eps2, d1, d2, V, communicator=MPI.COMM_WORLD):
+    def __init__(self, eps0, eps1, epsg = 1, eps2, d1, d2, voltage, communicator=MPI.COMM_WORLD):
```
Comment (sitangshugk95) [inline Adhesion/Interactions/ES_Coulomb.py:45]: Name changed to voltage
eps0 is now an input
eps1 is now an input (set to 1 as default, for air gap)

**Review by sitangshugk95 — COMMENTED** [review]
- Made changes to the code defining electrostatic potential as suggested (in the potential defn file)

- Added test to take input parameters consistent with human finger touching a haptic touchscreen; no direct results/plots exist for comparison, but can be qualitatively compared against another paper (both cited in comments in the test file)

File: Adhesion/Interactions/ES_Coulomb.py (hunk 1 above)
Comment (sannant) [reply Adhesion/Interactions/ES_Coulomb.py:2]: We have an automatic script doing this. 
run `sh maintenance/update_license_headers.sh` from the project main directory once everything else is done with the pull request.

File: test/Interactions/test_interactions.py [hunk 8]
```diff
@@ -1074,3 +1075,19 @@ def test_morse():
     range = 7.56
     pot = Morse(w, range)
     assert pot.evaluate(0)[0] == - w
+    
+def test_electrostatic_coulomb():
```
Comment (sannant) [inline test/Interactions/test_interactions.py:1079]: The test will not fail if the output of the function is wrong. 

1. add an assertion that the values are correct (e.g. extract some of the values from their plot and compare to them, doesn't need to be super accurate). 

2. wrap the plotting into an `if False:` so that the plot will not show up in normal testing sessions (would just be here for debugging purposes)

**Review by sannant — CHANGES_REQUESTED** [review]

File: test/Interactions/test_interactions.py (hunk 8 above)
Comment (sannant) [reply test/Interactions/test_interactions.py:1079]: Ah sorry just read the comment in the PR. if there is no such data then leave it. But wouldn't fig. 3 do part of the job ?

**Comment by sannant** [comment]
Can you actually see the test results or do we need to give you some privileges for that ? 
Do you know how to implement the flake8 conventions ? 
Do you know how to run tests locally ? 

There are very  convenient ways to take care of that in PyCharm

File: test/Interactions/test_interactions.py (hunk 8 above)
Comment (sitangshugk95) [reply test/Interactions/test_interactions.py:1079]: 1. Took test values and plotted, they look exactly as Fig. 3 now after non-dimensionalization
2. Plot wrapped in if False

**Comment by sannant** [comment]
In future, please follow our contributing guidelines concerning the  commit names. 

https://contactengineering.github.io/SurfaceTopography/contributing.html

File: examples/.ipynb_checkpoints/CMC_el_pl_adh_powerlaw-checkpoint.ipynb
Comment (sannant) [inline examples/.ipynb_checkpoints/CMC_el_pl_adh_powerlaw-checkpoint.ipynb:1]: Please do not add ipynb checkpoints !

**Comment by sannant** [comment]
Did you accidentally add all these notebooks ? 

Or what is their purpose ?

File: test/Interactions/ES_C_test.py
Comment (sannant) [inline test/Interactions/ES_C_test.py:1]: begin filename with test_ if it should be a test

File: test/Interactions/ES_C_test.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+# SOFTWARE.
+#
+"""
+Created on Wed Feb 22 22:52:48 2023
+
+@author: sitangshugk95
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn.metrics import mean_squared_error
```
Comment (sannant) [inline test/Interactions/ES_C_test.py:31]: please do not rely on new dependency to compute a mean squared error !
numpy or scipy is enough for that

File: test/Interactions/test_interactions.py [hunk 10]
```diff
@@ -1074,3 +1075,31 @@ def test_morse():
...
+    r = np.linspace(0*h0,5*h0,1000)
+    p,f,c = es_pot.evaluate(r)
+    x = r/h0
+    y = dV/sig0	    
+    indices = [0,200, 400, 600, 800, 1000]; 
+    y_test = [] #values calculated based on given input
+    for i in indices:
+    y_test.append(y[i])
+    y_true = [1.0, 0.25, 0.12, 0.625, 0.04, 0.028]; #true values from paper
+    error = mean_squared_error(y_true, y_test)
```
Comment (sannant) [inline test/Interactions/test_interactions.py:1100]: same as before concerning dependency

**Review by sannant — CHANGES_REQUESTED** [review]
Please remove or explain all the additional example files. 
We can also setup a zoom meeting for that

File: test/Interactions/ES_C_test.py (hunk 9 above)
Comment (sitangshugk95) [reply test/Interactions/ES_C_test.py]: > Please remove or explain all the additional example files. We can also setup a zoom meeting for that

They were added in error - all unwanted files have been deleted.

**Comment by sitangshugk95** [comment]
> And please make sure the tests pass before asking for a review.
> 
> Let us know if you need help to understand how to make these tests work

Yes, it would be helpful to know how to test these runs locally (instead of pushing it every time), and if there are any tests that are non-mandatory, or if ALL tests need to pass. Let me know if this would be more convenient on zoom, and I can email and coordinate with you.

**Comment by sitangshugk95** [comment]
I have debugged all remaining issues, and 10 out of the 11 tests were successful. 
However, the "Publish source package and wheel/build" test is failing, and I am not being able to understand what might be causing this and how to fix it. 
Screenshot from 2023-04-12 12-03-20

Could either of you help me out with this one @sannant @pastewka ?

**Comment by pastewka** [comment]
Thanks for your contribution @sitangshugk95 - I'm merging and will figure out problems in amster branch
